fix(helm): Set template context inside tpl function to outer function.

eicnix opened the pull request:
docs(helm): Added documentation about tpl function

Fix for https://github.com/kubernetes/helm/issues/3054

The Template Name of BasePath of the outer template which calls the tpl function will now be set during the execution of the tpl function.

eicnix committed:
fix(helm): Set template context inside tpl function to outer function.
docs(helm): Added documentation about tpl function

eicnix committed:
Fixed warning for missing formating parameter inside error message.

mparry commented:
Hi - I was looking at this problem locally a bit today, out of curiosity. The issue seemed to me to be where it is injecting `.Template` into the values during the `render()` call:

```
vals := tpls[file].vals
vals["Template"] = map[string]interface{}{"Name": file, "BasePath": tpls[file].basePath}
```

The `vals` structure here is the real global values, and so the real `.Values.Template` gets overwritten as soon as `tpl` is called. Wouldn't it be better to just avoid this injection altogether when `render()` is called by `tpl`, rather than making the dummy template in `tpl` pretend that it is the real top level one?

For example:

```
if _, ok := vals["Template"]; !ok {
	vals["Template"] = map[string]interface{}{"Name": file, "BasePath": tpls[file].basePath}
}
```

(Or else perhaps an explicit parameter to `render()`.)

eicnix commented:
I get your point and it should be decided if running `tpl` should be treated as a sub rendering processes or an independent rendering process. Based on this decision the template variable should either point to the outer template or to a `tpl` template.

bacongobbler commented on pkg/engine/engine.go:
```diff
@@ -166,19 +166,32 @@ func (e *Engine) alterFuncMap(t *template.Template) template.FuncMap {
+			basePath: basePath.(string),
 		}
 
+		println(vals.Table)
```
this needs to be removed

eicnix replied:
Debug statement removed

bacongobbler reviewed (changes requested):
can we also add unit tests here to ensure this won't break again? Thanks!

bacongobbler commented:
also looks like we need a rebase. Thanks!

eicnix committed:
fix(helm): Apply PR comments for tpl template name fix
Modified existing unit test to verify the changed behavior.
Removed debug print.

eicnix commented:
Hi @bacongobbler,

I made the changes you requested:
* removed debug statement
* changed existing unit test to include this specific behaviour
* rebased branch
